google_guest_agent/cfg/cfg.go

```diff
@@ -0,0 +1,299 @@
+	NetworkEnabled   bool   `ini:"network_enabled,omitempty"`
+	OptimizeLocalSSD bool   `ini:"optimize_local_ssd,omitempty"`
+	SetBotoConfig    bool   `ini:"set_boto_config,omitempty"`
+	SetHostKeys      bool   `ini:"set_host_keys,omitempty"`
+	SetMultiqueue    bool   `ini:"set_multiqueue,omitempty"`
+}
+
+// MetadataScripts contains the configurations of MetadataScripts section.
+type MetadataScripts struct {
+	DefaultShell      string `ini:"default_shell,omitempty"`
+	RunDir            string `ini:"run_dir,omitempty"`
+	Shutdown          bool   `ini:"shutdown,omitempty"`
+	ShutdownWindows   bool   `ini:"shutdown-windows,omitempty"`
+	Startup           bool   `ini:"startup,omitempty"`
+	StartupWindows    bool   `ini:"startup-windows,omitempty"`
+	SysprepSpecialize bool   `ini:"sysprep_specialize,omitempty"`
+}
+
+// NetworkInterfaces contains the configurations of NetworkInterfaces section.
+type NetworkInterfaces struct {
+	DHCPCommand  string `ini:"dhcp_command,omitempty"`
+	IPForwarding bool   `ini:"ip_forwarding,omitempty"`
+	Setup        bool   `ini:"setup,omitempty"`
+}
+
+// Snapshots contains the configurations of Snapshots section.
+type Snapshots struct {
+	Enabled             bool   `ini:"enabled,omitempty"`
+	SnapshotServiceIP   string `ini:"snapshot_service_ip,omitempty"`
+	SnapshotServicePort int    `ini:"snapshot_service_port,omitempty"`
+	TimeoutInSeconds    int    `ini:"timeout_in_seconds,omitempty"`
+}
+
+// Unstable contains the configurations of Unstable section. No long term stability or support
+// is guaranteed for configurations defined in the Unstable section. By default all flags defined
+// in this section is disabled and is intended to isolate under development features.
+type Unstable struct {
+}
+
+// WSFC contains the configurations of WSFC section.
+type WSFC struct {
+	Addresses string `ini:"addresses,omitempty"`
+	Enable    bool   `ini:"enable,omitempty"`
+	Port      string `ini:"port,omitempty"`
+}
+
+func defaultConfigFile(osName string) string {
+	if osName == "windows" {
+		return winConfigPath
+	}
+	return unixConfigPath
+}
+
+func defaultDataSources() []interface{} {
+	configFile := defaultConfigFile(runtime.GOOS)
+	return []interface{}{
+		[]byte(defaultConfig),
+		configFile,
+		configFile + ".distro",
+		configFile + ",template",
+	}
+}
+
+// Load loads default configuration and the configuration from default config files.
+func Load() error {
+	opts := ini.LoadOptions{
+		Loose:       true,
+		Insensitive: true,
+	}
+
+	sources := dataSources()
+	cfg, err := ini.LoadSources(opts, sources[0], sources[1:]...)
+	if err != nil {
+		return fmt.Errorf("failed to load configuration: %+v", err)
+	}
+
+	sections := new(Sections)
+	if err := cfg.MapTo(sections); err != nil {
+		return fmt.Errorf("failed to map configuration to object: %+v", err)
+	}
+
+	instance = sections
+	return nil
+}
+
+// Get returns the configuration's instance previously loaded with Load().
+func Get() (*Sections, error) {
```

Comment:
I wanted to make it very distinctive to what each call is doing, in guest-agent ideally we should have only one call to `Load()` and handled that as early init code. If we can't load defaults + user files we should abort - as a non recoverable error.

Reply:
It can follow singleton pattern, load & return on first call and then reuse. That way there are no concerns about ensuring its loaded as early as possible.

Reply:
My point is handling the error condition for the two operations differently (different meaning) and in different stages of to the code execution.